src/main/java/jenkins/plugins/git/GitRepoSizeEstimator.java

```diff
@@ -0,0 +1,168 @@
+        GitTool tool = GitUtils.resolveGitTool(gitImplementation, jenkins, null, TaskListener.NULL);
+        if (tool != null) {
+            gitTool = tool.getGitExe();
+        }
+    }
+
+    /**
+     * Determine and estimate the size of a .git cached directory
+     * @param source: Use a {@link AbstractGitSCMSource} to access a cached Jenkins directory, we do not lock it.
+     * @return useCache
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    private boolean setSizeFromCache(@NonNull AbstractGitSCMSource source) throws IOException, InterruptedException {
+        boolean useCache;
+        String cacheEntry = source.getCacheEntry();
+        File cacheDir = AbstractGitSCMSource.getCacheDir(cacheEntry);
+        if (cacheDir != null) {
+            Git git = Git.with(TaskListener.NULL, new EnvVars(EnvVars.masterEnvVars)).in(cacheDir).using("git");
+            GitClient client = git.getClient();
+            if (!client.hasGitRepo()) {
+                useCache = false;
+            } else {
+                useCache = true;
+                sizeOfRepo = FileUtils.sizeOfDirectory(cacheDir);
+                sizeOfRepo = (sizeOfRepo/1000); // Conversion from Bytes to Kilo Bytes
+            }
+        } else {
+            useCache = false;
+        }
+        return useCache;
+    }
+
+    /**
+     * Recommend a git implementation on the basis of the given size of a repository
+     * @param sizeOfRepo: Size of a repository (in KiBs)
+     * @return a git implementation, "git" or "jgit"
+     */
+    private String determineSwitchOnSize(Long sizeOfRepo) {
+        if (sizeOfRepo != 0L) {
+            if (sizeOfRepo >= SIZE_TO_SWITCH) {
+                return "git";
+            } else {
+                return "jgit";
+            }
+        }
+        return "NONE";
+    }
+
+    /**
+     * Other plugins can estimate the size of repository using this extension point
+     * The size is assumed to be in KiBs
+     */
+    public static abstract class RepositorySizeAPI implements ExtensionPoint {
+
+        public abstract Long getSizeOfRepository(String remote);
+
+        public static ExtensionList<RepositorySizeAPI> all() {
+            Jenkins jenkins = Jenkins.getInstanceOrNull();
+            if (jenkins == null) {
+                return null;
+            }
+            return jenkins.getExtensionList(RepositorySizeAPI.class);
+        }
+    }
+
+    private boolean setSizeFromAPI(String repoUrl) {
+        for (RepositorySizeAPI r: Objects.requireNonNull(RepositorySizeAPI.all())) {
+            if (r != null) {
+                sizeOfRepo = r.getSizeOfRepository(repoUrl);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Recommend git tool to be used by the git client
+     * @return git implementation recommendation in the form of a string
+     */
+    public String getGitTool() {
+        return gitTool;
+    }
```

Comment:
Wondering if it should return the GitTool object. @MarkEWaite what do you think?

Reply:
The reason for returning a `String` is this portion of code in AbstractGitSCMSource:
```
GitTool tool = resolveGitTool(context.gitTool(), listener);
            if (tool != null) {
                git.using(tool.getGitExe());
            }
            GitClient client = git.getClient();
```
We create a git client using the `git.using(tool.getGitExe())`. If we instantiate a GitRepoSizeEstimator just before it, it can provided a `String recommendedGitExe` to `git.using()` to create the client with the optimal implementation.